src/merkle/merkle-car-factory.ts

```diff
@@ -0,0 +1,90 @@
+import type { DiagnosticsLogger } from '@ceramicnetwork/common'
+import { CARFactory, CAR } from 'cartonne'
+import * as DAG_JOSE from 'dag-jose'
+import {
+  IpfsLeafCompare,
+  MerkleTreeFactory,
+  IpfsMerge,
+  type MerkleTree,
+  type Node,
+  type PathDirection,
+  type CIDHolder,
+  type TreeMetadata,
+} from '@ceramicnetwork/anchor-utils'
+import { BloomMetadata } from './bloom-metadata.js'
+import type { Candidate } from '../services/candidate.js'
+import type { CID } from 'multiformats/cid'
+
+const carFactory = new CARFactory()
+carFactory.codecs.add(DAG_JOSE)
+
+export class CARIpfsService {
+  readonly car: CAR
+  constructor(car?: CAR) {
+    this.car = car || carFactory.build()
+  }
+
+  async storeRecord(record: any): Promise<CID> {
+    return this.car.put(record)
+  }
+}
+
+export type MerkleCarFactoryResult = {
+  tree: MerkleTree<CIDHolder, Candidate, TreeMetadata>
+  car: CAR
+}
+
+export interface IMerkleTree<TData, TLeaf extends TData, TMetadata> {
+  readonly root: Node<TData>
+  readonly leafNodes: Array<Node<TLeaf>>
+  readonly metadata: TMetadata | null
+  getProof(elemIndex: number): Array<Node<TData>>
+  getDirectPathFromRoot(elemIndex: number): PathDirection[]
+  verifyProof(proof: Node<TData>[], element: TData): Promise<boolean>
+}
+
+export class MerkleCAR implements IMerkleTree<CIDHolder, Candidate, TreeMetadata> {
```

Comment:
why do we need this class?  It doesn't really seem to do anything but associate a CAR with a MerkleTree - all the functionality though is a direct pass-through to the underlying merkle tree.  Would it be simpler to just pass around a struct with both the tree and the car file?

Reply:
Nevermind, I see now.  It overrides the IPFSService to ensure that as the tree is being built each intermediate node in the tree is added to the CAR file.  That behavior is a bit subtle, some more comments throughout this file could help.